crates/sui-core/src/transaction_manager.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
+
+use sui_storage::node_sync_store::NodeSyncStore;
+use sui_types::{
+    base_types::TransactionDigest, committee::EpochId, error::SuiResult,
+    messages::VerifiedCertificate,
+};
+use tokio::sync::mpsc::UnboundedSender;
+use tracing::error;
+
+use crate::authority::{authority_store::ObjectKey, AuthorityMetrics, AuthorityStore};
+
+/// TransactionManager is responsible for managing pending certificates and publishes a stream
+/// of certificates ready to be executed. It works together with AuthorityState for receiving
+/// pending certificates, and getting notified about committed objects. Executing driver
+/// subscribes to the stream of ready certificates published by the TransactionManager, and can
+/// execute them in parallel.
+/// TODO: use TransactionManager for fullnode.
+pub(crate) struct TransactionManager {
+    authority_store: Arc<AuthorityStore>,
+    node_sync_store: Arc<NodeSyncStore>,
+    missing_inputs: BTreeMap<ObjectKey, (EpochId, TransactionDigest)>,
+    pending_certificates: BTreeMap<(EpochId, TransactionDigest), BTreeSet<ObjectKey>>,
+    tx_ready_certificates: UnboundedSender<VerifiedCertificate>,
+    metrics: Arc<AuthorityMetrics>,
+}
+
+impl TransactionManager {
+    /// If a node restarts, transaction manager recovers in-memory data from pending certificates and
+    /// other persistent data.
+    pub(crate) fn new(
+        authority_store: Arc<AuthorityStore>,
+        node_sync_store: Arc<NodeSyncStore>,
+        tx_ready_certificates: UnboundedSender<VerifiedCertificate>,
+        metrics: Arc<AuthorityMetrics>,
+    ) -> TransactionManager {
+        let mut transaction_manager = TransactionManager {
+            authority_store,
+            node_sync_store: node_sync_store.clone(),
+            metrics,
+            missing_inputs: BTreeMap::new(),
+            pending_certificates: BTreeMap::new(),
+            tx_ready_certificates,
+        };
+        transaction_manager
+            .enqueue(node_sync_store.all_pending_certs().unwrap())
+            .expect("Initialize TransactionManager with pending certificates failed.");
+        transaction_manager
+    }
+
+    /// Enqueues a certificate into TransactionManager. Once all of its input objects are available
+    /// locally, it will be published.
+    /// TODO: turn this function async, and wait for inflight certificates below threshold.
```

Comment:
I don't think this TODO is possible - if we bound the number of inflight certificates, we may deadlock, because we may be awaiting the completion of a cert whose dependencies we are trying to enqueue.